### PR TITLE
V4: Cache overhaul, gql_link integration, and more

### LINKED ISSUE #608: Roadmap for v4
There are a handful of open PRs and issues that I think should be parceled together for the next major version:
- [X] #564 & #565 (outsource links to gql)
- [x]  many caching fixes:
  - [x] #471 (outsource normalization)
  - [x] #423 (the cache is kinda broken)
- [x] many of the items enumerated by @klavs in https://github.com/zino-app/graphql-flutter/issues/563#issuecomment-587806331

The **main question** is should we make a `v4` branch or something to avoid every merge auto-publishing a breaking change? **Or**, should we rework CI such that `master` is the main development branch, and have `master, beta, stable`? **Or** should we just temporarily disable CI and then re-enable after achieving a stable API?

@HofmannZ @eusdima @mainawycliffe

### PR DESCRIPTION
**Pull request by micimize**
## [`4.0.0-alpha.11`] is published!

See the [v4 changelog] - we're offloading links onto the [gql_link system], and now have a [normalize]-backed cache (heavily based on [ferry]'s cache).

> v4 aims to solve a number of sore spots, particularly with caching, largely by leveraging libraries from the https://github.com/gql-dart ecosystem. There has also been a concerted effort to add more API docstrings to the code base.

Will close #608, and thus by extension #176, #471, #646, #300, and probably many other deep-seeded cache issues

If you are not using **file uploads** or the **`PersistedQueriesLink`**, please give it a try and [report any issues] you encounter, or close/comment any issues it solves for you. If you didn't have any, or just want to give feedback, let us know in the [discord](https://discord.gg/tXTtBfC) `#feedback` channel!

If you want and can afford to **support ongoing development**, you can [❤️ sponsor me](https://github.com/sponsors/micimize/) or contribute to the [👐 open collective](https://opencollective.com/graphql-flutter).

[v4 changelog]: https://github.com/zino-app/graphql-flutter/blob/modularization/changelog-v3-v4.md
[ferry]: https://github.com/gql-dart/ferry/
[gql_link system]: https://github.com/gql-dart/gql/tree/master/links/gql_link
[`4.0.0-alpha.11`]: https://pub.dev/packages/graphql/versions/4.0.0-alpha.11
[report any issues]: https://github.com/zino-app/graphql-flutter/issues/new?assignees=micimize&labels=v4&template=v4-issue.md&title=
[normalize]: https://pub.dev/packages/normalize

### CODE REVIEW

**Comment by teknolojia** [comment]
@micimize our startup needs session so badly right now. If you could help bring that thing soon enough we promise to sponsor this project as much as we can if you help us reach MVP.

**Comment by micimize** [comment]
@HofmannZ can you add me to the pub publisher account? I want to manually publish `alpha` builds for user testing before merging into `beta`

**Comment by HofmannZ** [comment]
@micimize Sure! Will have a look at this PR when I find some time. It's a beast!

**Comment by RodolfoSilva** [comment]
`QueryResult` is `null` when the `Subscription` component is mountend in the first time.

File: examples/starwars/pubspec.yaml [hunk 1]
```diff
@@ -8,15 +8,14 @@ dependencies:
     path: ../../packages/graphql_flutter
   graphql:
     path: ../../packages/graphql
-  universal_platform: ^0.1.3
```
Comment (RodolfoSilva) [inline examples/starwars/pubspec.yaml:11]: What's the difference between these definitions?

Comment (micimize) [reply examples/starwars/pubspec.yaml:11]: @RodolfoSilva just autoformatting from vscode, prob caused by the comment

**Comment by aaahrens** [comment]
does this also include the @client directive?

**Comment by rwrz** [comment]
I'm using this, any help needed?

**Comment by micimize** [comment]
@aaahrens no – `@client` looks fairly involved so probably won't be implementing it anytime soon

@rwrz If you want to take a crack at one of the the few [outstanding issues](https://github.com/zino-app/graphql-flutter/issues?q=is%3Aissue+is%3Aopen+label%3Av4) it'd be appreciated! Otherwise, any feedback/thoughts you have are welcome in the [discord](https://discord.gg/tXTtBfC) `#feedback` channel.

Will probably be bumping v4 into beta after resolving the outstanding issues and getting `gql_http_link: 0.3.0` published. Haven't had much time to work on said issues recently though

**Comment by micimize** [comment]
With the revival of the old `WebSocketLink` and the last round of bug fixes, this is almost ready to move in to `beta`. I'm going to try and set aside some time in the next week or two for some cleanup, get CI working again, and publish.

Then once we've got a beta, I'll merge and publish the request method context in https://github.com/gql-dart/gql/pull/123, and break the dynamic serialization context into a custom serializer provided by  `graphql` for use with `apq` and merge in #699 

Then a round of more rigorous documentation

**Comment by micimize** [comment]
[`4.0.0-alpha.11`](https://pub.dev/packages/graphql/versions/4.0.0-alpha.11) is out with some regression fixes and a breaking change to the fragment read/write api from `alpha.9`.

As there has been 1 major regression found/fixed this last week, I'm going to leave the alpha out for another week or two before upgrading to `beta` in case any more are found.

Also need to give `pub.dev` time to generate API Docs for linking from the readmes, which apparently takes an insanely long amount of time.

**Pull request merged**